src/split_single_dyn_dim.cpp

```diff
@@ -36,36 +36,79 @@ inline namespace MIGRAPHX_INLINE_NS {
+has_one_unique_dyn_dim(const std::unordered_map<std::string, shape>& param_shapes)
 {
-    // True if parameters contain exactly one dynamic shape with exactly one non-fixed
-    // dynamic_dimension.
+    if(param_shapes.empty())
+    {
+        return std::nullopt;
+    }
     auto is_dynamic = [](const auto& p) { return p.second.dynamic(); };
-    auto ps_it      = std::find_if(param_shapes.begin(), param_shapes.end(), is_dynamic);
-    if(ps_it == param_shapes.end())
+    std::vector<std::decay_t<decltype(param_shapes)>::value_type> dyn_params{};
+    std::copy_if(
+        param_shapes.begin(), param_shapes.end(), std::back_inserter(dyn_params), is_dynamic);
+    if(dyn_params.empty())
         return std::nullopt;
-    // Check if there is a second dynamic parameter
-    if(std::any_of(std::next(ps_it), param_shapes.end(), is_dynamic))
+    std::vector<dynamic_dimensions_check> ret{};
+    // get non-fixed dynamic_dimension from all parameters
+    for(const auto& param : dyn_params)
+    {
+        const auto& dds   = param.second.dyn_dims();
+        int num_non_fixed = 0;
+        for(const auto& dd : dds)
+        {
+            if(not dd.is_fixed())
+            {
+                num_non_fixed += 1;
+                // catch more than one non-fixed dynamic_dimension
+                if(num_non_fixed > 1)
+                {
+                    return std::nullopt;
+                }
+                ret.push_back(dynamic_dimensions_check{param.first, dd});
+            }
+        }
+    }
+    if(ret.empty())
+    {
         return std::nullopt;
-    const auto& dds = ps_it->second.dyn_dims();
+    }
+    // check all the same dynamic_dimension
+    bool same_dd =
+        std::all_of(ret.begin() + 1, ret.end(), [&](auto ddc) { return ddc.dd == ret.at(0).dd; });
+    if(same_dd)
+    {
+        return ret;
+    }
+    return std::nullopt;
+}
 
-    auto is_non_fixed = [](const auto& dd) { return not dd.is_fixed(); };
-    auto dds_it       = std::find_if(dds.begin(), dds.end(), is_non_fixed);
-    if(dds_it == dds.end())
-        return std::nullopt;
-    // Check if there is a second non-fixed dynamic_dimension
-    if(std::any_of(std::next(dds_it), dds.end(), is_non_fixed))
-        return std::nullopt;
-    return dynamic_dimensions_check{ps_it->first,
-                                    static_cast<std::size_t>(std::distance(dds.begin(), dds_it)),
-                                    dds_it->min,
-                                    dds_it->max};
+/**
+ * Check the parameters in std::vector<dynamic_dimensions_check> object to see if any of the
+ * parameters outputs to a select_module operator.
+ */
+bool any_sm_next(const_module_ref mm, const std::vector<dynamic_dimensions_check>& ddcs)
+{
+    for(const auto& ddc : ddcs)
+    {
+        auto p_outputs  = mm->get_parameter(ddc.dyn_param_str)->outputs();
+        bool is_sm_next = std::any_of(p_outputs.cbegin(), p_outputs.cend(), [](auto ins) {
```

Comment:
You can just do `return std::any_of`

Reply:
Cannot, need to continue the loop if a select_module is not found. I could move the `std::any_of` call into the `if`, but that looks pretty messy to me.